src/api.v2/helpers/pipeline/handleQCResponse.js

```diff
@@ -0,0 +1,147 @@
+hookRunner.registerAll([sendNotification]);
+hookRunner.register('configureEmbedding', [cleanupPods]);
+
+const getS3Output = async (message) => {
+  const { output: { bucket, key } } = message;
+  const s3 = new AWS.S3();
+  const outputObject = await s3.getObject(
+    {
+      Bucket: bucket,
+      Key: key,
+    },
+  ).promise();
+
+  return JSON.parse(outputObject.Body.toString());
+};
+
+const updatePlotData = async (taskName, experimentId, output) => {
+  if (output.plotDataKeys) {
+    const plotConfigUploads = Object.entries(output.plotDataKeys)
+      .map(async ([plotUuid, s3DataKey]) => (
+        await new Plot().upsert(
+          { id: plotUuid, experiment_id: experimentId },
+          { s3_data_key: s3DataKey },
+        )
+      ));
+
+    logger.log('Uploading plotData for', taskName, 'to DynamoDB');
+
+    // Promise.all stops if it encounters errors.
+    // This handles errors so that error in one upload does not stop others
+    await Promise.all(
+      plotConfigUploads.map((p) => p.catch((e) => logger.error(e))),
+    );
+  }
+};
+
+const updateProcessingConfig = async (taskName, experimentId, output, sampleUuid) => {
+  // TODO the processing config validation was not being enforced because the 'anyOf' requirement
+  // was not being correctly applied. This needs to be refactored together with the
+  // pipeline and ideally while unifying the qc & gem2s responses.
+  // await validateRequest(output, 'ProcessingConfigBodies.v1.yaml');
+
+  const experiment = new Experiment();
+
+  const {
+    processingConfig: previousConfig,
+  } = await experiment.findById(experimentId).first();
+
+  if (sampleUuid !== '') {
+    const { auto } = output.config;
+
+    // This is a temporary fix to save defaultFilterSettings calculated in the QC pipeline
+    // to patch for old experiments with hardcoded defaultFilterSettings.
+    // Remove this once we're done migrating to the new experiment schema with
+    // defaultFilterSettings
+    const sampleOutput = output;
+
+    if (auto) sampleOutput.config.defaultFilterSettings = output.config.filterSettings;
+
+    await experiment.updateProcessingConfig(experimentId, [
+      {
+        name: taskName,
+        body: {
+          ...previousConfig[taskName],
+          [sampleUuid]: { ...sampleOutput.config },
+        },
+      },
+    ]);
+  } else {
+    await experiment.updateProcessingConfig(experimentId, [
+      {
+        name: taskName,
+        body: output.config,
+      },
+    ]);
+  }
+};
+
+const sendUpdateToSubscribed = async (experimentId, message, output, error, io) => {
+  const statusRes = await getPipelineStatus(experimentId, constants.QC_PROCESS_NAME);
+  const statusResToSend = { pipeline: statusRes[constants.QC_PROCESS_NAME] };
+
+  // Concatenate into a proper response.
+  const response = {
+    ...message,
+    status: statusResToSend,
+    type: constants.QC_PROCESS_NAME,
+  };
+
+  if (output !== null) {
```

Comment:
can `output` be null? Based on my code reading, if `output` is null, then `if (output.plotDataKeys)` inside function `updatePlotData` will error out. Or am I missing something?

Reply:
I could make it `output?.plotDataKeys`